ambari-server/src/main/python/ambari_server/setupSecurity.py

```diff
@@ -638,63 +647,98 @@ def setup_ambari_krb5_jaas(options):
 
 
 class LdapPropTemplate:
-  def __init__(self, properties, i_option, i_prop_name, i_prop_val_pattern, i_prompt_regex, i_allow_empty_prompt, i_prop_name_default=None):
+  def __init__(self, properties, i_option, i_prop_name, i_prop_val_pattern, i_prompt_regex, i_allow_empty_prompt, i_prop_default=None):
     self.prop_name = i_prop_name
     self.option = i_option
-    self.ldap_prop_value = get_value_from_properties(properties, i_prop_name, i_prop_name_default)
-    self.ldap_prop_val_prompt = format_prop_val_prompt(i_prop_val_pattern, self.ldap_prop_value)
+    stored_value = get_value_from_properties(properties, i_prop_name)
+    self.default_value = LdapDefault(stored_value) if stored_value else i_prop_default
+    self.prompt_pattern = i_prop_val_pattern
     self.prompt_regex = i_prompt_regex
     self.allow_empty_prompt = i_allow_empty_prompt
 
+  def get_default_value(self, ldap_type):
+    return self.default_value.get_default_value(ldap_type) if self.default_value else None
+
+  def get_prompt_text(self, ldap_type):
+    default_value = self.get_default_value(ldap_type)
+    return format_prop_val_prompt(self.prompt_pattern, default_value)
+
+  def get_input(self, ldap_type):
+    default_value = self.get_default_value(ldap_type)
+    return get_validated_string_input(self.get_prompt_text(ldap_type),
+                                      default_value, self.prompt_regex,
+                                       "Invalid characters in the input!", False, self.allow_empty_prompt,
+                                       answer = self.option)
+
+  def should_query_ldap_type(self):
+    return not self.allow_empty_prompt and not self.option and self.default_value and self.default_value.depends_on_ldap_type()
+
+class LdapDefault:
+  def __init__(self, value):
+    self.default_value = value
+
+  def get_default_value(self, ldap_type):
+    return self.default_value
+
+  def depends_on_ldap_type(self):
+    return False
+
+
+class LdapDefaultMap(LdapDefault):
+  def __init__(self, value_map):
+    LdapDefault.__init__(self, None)
+    self.default_value_map = value_map
+
+  def get_default_value(self, ldap_type):
+    return self.default_value_map[ldap_type] if self.default_value_map and ldap_type in self.default_value_map else None
+
+  def depends_on_ldap_type(self):
+    return True
+
 def format_prop_val_prompt(prop_prompt_pattern, prop_default_value):
   default_value = get_prompt_default(prop_default_value)
   return prop_prompt_pattern.format((" " + default_value) if default_value is not None and default_value != "" else "")
 
 @OsFamilyFuncImpl(OSConst.WINSRV_FAMILY)
```

Comment:
Should we get rid of these WINSRV_FAMILY functions? I don't think windows is supported anymore.